File: 1297-maximum-number-of-balloons/1297-maximum-number-of-balloons.py
class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        dct = {'b':0, 'a': 0, 'l': 0, 'o' : 0, 'n': 0}
        mini = sys.maxsize
        for t in text:
            if t in dct:
                dct[t] += 1
        
        for (key, val) in dct.items():
            if key == 'l' or key == 'o':
                val //= 2
            else:
                val
            mini = min(mini, val)
        return mini

    # def maxNumberOfBalloons(self, text: str) -> int:
    #     b = text.count("b")
    #     a = text.count("a")
    #     l = text.count("l")
    #     o = text.count("o")
    #     n = text.count("n")

    #     return min(b, a, l // 2, o // 2, n)


        

File: 1297-maximum-number-of-balloons/1297-maximum-number-of-balloons.rs

use std::collections::HashMap;
impl Solution {
    pub fn max_number_of_balloons(text: String) -> i32 {
        let mut instances : HashMap<char, i32> = HashMap::new();
        instances.insert('b', 0);
        instances.insert('a', 0);
        instances.insert('l', 0);
        instances.insert('o', 0);
        instances.insert('n', 0);
        for t in text.chars(){
            if instances.contains_key(&t){
                *instances.entry(t).or_insert(0) += 1;
            }
        }

        let mut word = 0;
        let mut mini = i32::MAX;
        let size = instances.len();
        for (&key, &val) in &instances{
            let true_val = if key == 'l' || key == 'o' {
                val / 2
            } else {
                val
            };
            mini = mini.min(true_val);
        }
        if mini == i32::MAX{
            -1
        }else{
            mini
        }
    }
}

// Character to Index Conversion:
// ch as u8: Converts the character ch to its ASCII byte value.
// b'a': Represents the ASCII value of the character 'a'.
// ch as u8 - b'a': Calculates the zero-based index for the character, where 'a' maps to 0, 'b' maps to 1, and so on.
// as usize: Converts the resulting index to usize because array indexing in Rust requires usize type.

// pub fn max_number_of_balloons(text: String) -> i32 {
    //     let mut freq = [0; 26];
    //     for ch in text.chars() {
    //         freq[(ch as u8 - b'a') as usize] += 1;
    //     }

    //     let mut ans = 0;
    //     'out: loop {
    //         for ch in "balloon".chars() {
    //             let i = (ch as u8 - b'a') as usize;
    //             if freq[i] == 0 {
    //                 break 'out;
    //             }

    //             freq[i] -= 1;
    //         }
    //         ans += 1;
    //     }

    //     ans
    // }